사다리타기 첫 번째 구현

lette1394 opened the pull request:
사다리타기 첫 번째 구현 

리뷰 부탁 드립니다.

lette1394 committed:
initial commit

Add build.gradle
ADd Main class - 'hello world!' for test
Add .gitignore

lette1394 committed:
Add Main Test for standard input output

lette1394 committed:
Pass Main test

lette1394 committed:
Fix Main test glitches

lette1394 committed:
Add test for parsing initial setting inputs

사다리 결과가 나오기 전까지의 유저 입력을 파싱, 저장하는 클래스에 대한 테스트 작성

lette1394 committed:
Fix Main output lines

lette1394 committed:
Implements for SettingProcessor test

lette1394 committed:
Change test method name, Add description to them

더 의미있는 이름과 설명을 추가

lette1394 committed:
Change test method name, Add description to them for MainTest

더 의미있는 이름과 설명을 추가

lette1394 committed:
Change method name handle -> parse in Parser

lette1394 committed:
Merge Setting, Setting.Type

같은 일을 하는데 굳이 나눠놓은 느낌이라 합쳤다.

lette1394 committed:
Apply SettingProcessor to Main

SettingProcessor가 InputStream을 받는 것이 아니라 Scanner를 받음.
System.in 공유자원 문제가 발생

두 군데 이상에서 System.in을 가지고 Scanner 를 생성하면 인덱스가 꼬이는듯함.

lette1394 committed:
Add RendererTest

lette1394 committed:
Implements for Renderer (Cont`d)

Extract StringUtils

lette1394 committed:
Implements for Renderer random line

lette1394 committed:
Extract RenderPolicy from Renderer

어렵다어려워...

lette1394 commented:
**아직 구현 중**

lette1394 committed:
Apply Renderer to Main, MainTest

테스트 용이성을 위해 MainTest를 OneLine Render Policy로 정의했다.
또한 한글을 지원하기 까다로운 문제로, MainTest의 입출력을 모두 영문으로 재정의했다.
추후 Random하게 사다리를 그리는 것은 테스트가 어렵기 때문에, Render Policy에 대한 입출력을 정의한 뒤에 테스트를 진행해야 할 듯 하다.

lette1394 committed:
Change Constructor of Renderer to clear meaning

Renderer 가 사다리를 정책이 생성할때 제대로 보이지 않고,
따라서 불필요하게 Renderer.with()를 통해 호출해야 하는 것이 부자연스럽다.

그러면 Renderer가 하는 일이 없어진다.
Renderer를 사용하는 클라이언트가 생각하기에, Renderer에 RenderContext안에 들어가는
이것저것 설정해주는 정보들을 바로 넣어주는 것이 이상적이지 않은가?

lette1394 committed:
Add TODO comments

lette1394 committed:
Renderer now hold RenderContext

lette1394 committed:
Add SettingContextTest toSwap method

사다리를 그릴 때 각 row별로 swap이 될 때마다, 결과를 들고 있는 Context를 변경하기 위해
불변식으로 작성하기 위해 toSwap() 메서드를 구현. 이에 대한 테스트

lette1394 committed:
Add Implementation for SettingContextTest

Swap부분이 마음에 안 들긴 하지만 이쁘게 코드를 짜기 위해선...

lette1394 committed:
Add Test RandomRenderPolicyTest

랜덤을 구현하기 어려워서, RandomBoolean이라는 인터페이스를 파고,
Renderer가 이 인터페이스에서 가로줄을 그을지 말지 결정을 한다.

해당 RandomBoolean을 구현하는 인터페이스를 여러개 구현하는데,
여기서 내가 미리 정의한 PredefinedRandom 을 넣어주면 mocking이 되는 것처럼 테스트가 가능해진다.

lette1394 committed:
Add TODO list

lette1394 committed:
Add TODOs, and display names for test

lette1394 committed:
Fix order of input messages

lette1394 committed:
Add default value of maxHeight to SettingContext Builder

builder에서 명시적으로 maxHeight를 호출하지 않았을 경우.
그러면 다른 필드들도 해줘야 하는거 아니냐... -> 일단 나중에 하자

lette1394 committed:
Add Query part. QueryProcessor, QueryCommand, QueryResult, Stringifiable

쿼리에 대한 결과, 각 결과는 현재 String으로만 출력하기 때문에 인터페이스를 따로 팠음
QueryCommand는 유저의 요청별로 Command를 만들기 위해. 기 정의된 동작을 수행할 것임

QueryCommand의 생성하는 책임은 QueryProcessor에 있음

lette1394 committed:
Add makeResultWithNameAndResult in StringUtils

QueryResult 에 들어갈 String을 만들기 위해 로직 부분을 StringUtils로 따로 분리
이걸 다시 재사용할까? 하는 생각이 들었지만, 우선 테스트에서 재사용할 것이므로
분리하기로 결정

lette1394 committed:
Change QueryProcessorTest better to generify test arguments

파라메터 테스트를 활용해서 중복 코드를 제거. 다른 Renderer 테스트에도 적용해야함.

lette1394 committed:
Implements for QueryProcessorTest

QueryCommand를 미리 정의해놓고 enum안에서 재사용하는 방식으로 정의
갈수록 SettingContext라는 이름은 어울리지 않는다.

1. UserInputContext
2. LadderGameContext

1번이 아무래도 나아보인다.

lette1394 committed:
Add toString to value object

lette1394 committed:
Add IllegalQueryCommandException for test

lette1394 committed:
Add QueryArgument interface and some implements

어떤 QueryCommand는 추가적인 정보가 필요하다. 이를 위해 QueryArgument interface를
작성하고, QueryProcessor가 유저로부터 입력을 받아 적절한 QeuryArgument 구현체를 선택한다.

lette1394 committed:
Add more test for SHOW query command, exception with wrong command

DisplayName을 더욱 의미 있게 수정
IDE의 도움을 받기 위해서 MethodSource로 필드를 주었던 것을 하드코딩으로 변경
SHOW command에 대한 테스트를 작성하는데, MethodSource로 테스트에 필요한 정보를 모두 받음
command 종류가 더 추가되면 주석 처리된 테스트를 진행

lette1394 committed:
Add implements for additional QueryProcessor test

Logging 및 Exception을 위해서 QueryCommand에 toString을 추가.
이를 위해서 String 필드를 추가함

lette1394 committed:
Reformat RandomRenderPolicyTest using @Nested annotation to be more readable

lette1394 committed:
Add more test for Main

lette1394 committed:
Fix bugs in StringUtils

lette1394 committed:
Extract run method by RenderPolicy to test easily

lette1394 committed:
Rename QueryProcessorTest -> QueryCommandTest

내가 생각한 QueryProcessor는

1. 사용자의 입력을 받고
2. 적절한 QueryCommand를 생성해서
3. 각 Command에 맞는 결과를 요청하고
4. 결과를 반환하는

애였는데, QueryProcessor가 QueryCommand가 하는 일을 그대로 해버렸다.

lette1394 committed:
Edit QueryProcessorTest to fit changes

이제 커맨드는 앞에 '@'를 붙인다. -> MainTest 수정

lette1394 committed:
Apply QueryProcessor to Main

lette1394 committed:
Change to use inner method in Renderer

lette1394 committed:
Remove value() in QueryCommand, Add run() in QueryProcessor

enum 안에서 name()을 사용하도록 변경
run() 메서드는 기존 Main에 있던 것을 활용

lette1394 committed:
Add polymorphism to parse QueryCommand, QueryArgument

lette1394 committed:
Rename QueryArgument -> QueryCommandArgument

Add factory method in QueryResult
Add IllegalQueryCommandArgumentException
Move PredefinedRandom to test package (only use in test)

lette1394 committed:
Rename SettingContext -> GameContext

Context라는 이름이 마음에 안들긴 하지만...

lette1394 committed:
Rename SettingContextProcessor -> GameContextProcessor

Context라는 이름이 마음에 안들긴 하지만...

lette1394 committed:
Change default RenderContext

너비를 8에서 7로 수정

lette1394 committed:
Add StandardInputOutput

remove TODO

lette1394 committed:
Add IO interface that implements Input, Output

Add no argument println() method in StandardInputOut

lette1394 committed:
Apply IO to classes that use Scanner

lette1394 committed:
Remove unused codes

lette1394 committed:
Rename package setting -> game

lette1394 committed:
Move Participant class to game package

lette1394 committed:
Move Participant class to model package

lette1394 committed:
Add QueryProcessor exception test to test invalid user name

lette1394 committed:
Add unsupported command test for QueryProcessor (not implemented)

lette1394 committed:
Extract common method that used in RenderPolicies

lette1394 committed:
Add logic to constraint of rendering horizontal ladder line

이전에 사다리가 그려져있으면, 다음 라인에 연속으로 그리지 않는 로직을 추가
관련된 테스트도 수정

lette1394 committed:
Add test display name

lette1394 committed:
Fix typo

lette1394 commented:
1차 구현 완료

@JuhoKang @HeechanYang

lette1394 committed:
Remove TODO comments

JuhoKang commented on ladder/jaeeun-na-1/src/main/java/org/ooptraining/game/Setting.java:
```diff
@@ -0,0 +1,22 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum Setting implements Parser {
```
요쪽 `Setting`에서 사용하는 `Parser`인터페이스의 `parse`가 뭘하는지 잘 모르겠는 부분이 있음.

`parse`를 하면 `GameContext.Builder`에 뭔가가 set? 추가되는식인가 싶기도하고. 코드가 굉장히 복잡한데 일반적으로 `parse`하면 뭔가의 결과값을 리턴받을거라고 기대할텐데 여기 코드는 `void`리턴이기도 해서. 설명이 좀 필요할듯.

lette1394 replied:
우선 `Setting`이름과 `Parser`이름이 진짜 어울리지 않아서 헷갈렸을 것 같어
이 부분을 비교적 초반에 짠거라 적절한 이름을 떠올리는게 어려웠어

우선, 내 의도는 `GameContext`가 전체 게임에 대한 유저 입력을 들고 있게 하는 게 목적이었어
`GameContext`는 각 `Setting`에게 파싱 명령을 내려서 `GameContext.Builder`를 통해 온전한 `GameContext`를 만들 수 있게 하는 것까지 책임이 있어

각 `Parser`의 구현체는 유저로부터 어떻게 입력을 가져오는지를 아는 녀석들이고, 가져온 뒤에는 Builder에 추가하는 식. 미리 정의된 `MaxHeightParser`같은 놈들이 그 예시야

`Parser`안에서 파싱도하고 `GameContext.Builder`에 추가도 하니까 메서드가 하는 일이 두개여서 그런듯
이걸 분리해야겠어

JuhoKang commented on ladder/jaeeun-na-1/src/main/java/org/ooptraining/io/StandardInputOutput.java:
```diff
@@ -0,0 +1,29 @@
+import java.io.PrintStream;
+import java.util.Scanner;
+
+public class StandardInputOutput implements IO {
```
`IO`라는 중간단계와 `Input`, `Output`을 다중 상속?(implements)하게 하고, 이렇게 StandardInputOutput이라는 클래스를 정의한 이유가 있을까. 확장대비?

lette1394 replied:
그렇지 
파일에서 입력 받고 출력은 표준출력으로 내보낼 수도 있도록

lette1394 replied:
de9e847 3ba899e eadc0da 참고하면 좋을 듯해

JuhoKang commented on ladder/jaeeun-na-1/src/main/java/org/ooptraining/io/Stringifiable.java:
```diff
@@ -0,0 +1,5 @@
+package org.ooptraining.io;
+
+public interface Stringifiable {
```
`asString`과 `toString`은 어떤 차이가 있는지 궁금. 뉘앙스의 차이가 있는것 같긴한데. 주석으로라도 표현이 필요해 보임.
Stringify라고 생각하면 사실 JSON.stringify가 더 떠올라서 좀 다른것 같기도하고

lette1394 replied:
지금 하나 밖에 없어서 불필요하게 보이긴 한데, 
`io` 패키지 안에서 `asString`뿐 아니라 `asBinary`, `asStream<T>` 이런식으로 확장할 수 있도록 만들어 놓은 거야 

추후에 `QueryResult`가 저 인터페이스만 구현하면 `Stream<T>`만 입력으로 받는 `io`에 대응할 수 있도록

JuhoKang commented on ladder/jaeeun-na-1/src/main/java/org/ooptraining/model/Participant.java:
```diff
@@ -0,0 +1,17 @@
+    private final String name;
+    private final String result;
+
+    public Participant clone() {
```
Cloneable한 객체 정의가 쉽지 않은데 굳이 쓴 이유가 있을까 싶음.
사실 이부분은 나도 잘 모르는데 effective java 보면 조심스럽게 clone override하라고 하는데 잘 아는게 있으면 나도 좀 알려주십사

http://thefinestartist.com/effective-java/11

lette1394 replied:
자바 Clonable은 규약이 굉장히 허술한 것으로 알고 있어 
내 의도는 `deepCopy()`가 목적이었는데, 왜 `clone`을 정의했는지 모르겠네

`deepCopy()`로 수정해둘게

JuhoKang commented on ladder/jaeeun-na-1/src/main/java/org/ooptraining/query/QueryCommand.java:
```diff
@@ -0,0 +1,97 @@
+import static java.util.stream.Collectors.toMap;
+import static org.ooptraining.util.StringUtils.makeFullResultWithNameAndResult;
+
+public enum QueryCommand {
```
`QueryCommand` 관련 로직이 한눈에 보기 어려워서 굉장히 복잡한데, 혹시 참고한 코드나, 디자인 패턴같은게 있는지 있으면 읽는데 도움이 될것 같음.

lette1394 replied:
`public abstract QueryResult execute(final GameContext gameContext, final QueryCommandArgument queryCommandArgument);` 
요게 핵심이야

`GameContext` : 게임에 대한 유저 입력을 알고 있음
`QueryCommandArgument` : 쿼리가 실행되는데 필요한 추가 `argument`
-> `QueryCommand`는 미리 정의된 동작을 `argument`와 `context`에 따라서 실행한다.

현재 커맨드는 각각 `SHOW_ALL`, `SHOW`, `EXIT`가 있고,
하나 예를 들면 `SHOW_ALL`은 `argument`가 들어와도 무시하고, gameContext안에 있는 정보에 따라 결과를 만들어서 반환해 (`makeFullResultWithNameAndResult`)

JuhoKang commented on ladder/jaeeun-na-1/src/main/java/org/ooptraining/query/QueryProcessor.java:
```diff
@@ -0,0 +1,51 @@
+    public void run(final GameContext gameContext) {
+        try {
+            while (true) {
+                io.println("결과를 보고 싶은 사람은?");
```
보다 보면 느끼는데 정말 `System.out`을 `io`로 대체해서 많은 이익을 얻을수 있을까 싶은 느낌이 듬. 안드로이드로 옮겼다고 가정했을때, AndroidInputOutput만 만들면 사다리 게임이 옮겨질까?

lette1394 replied:
옮겨진다고 봐

안드로이드에서 유저 입력을 받는 Adaptor를 `Input`인터페이스를 구현하게 만들고, 
마찬가지로 유저 화면에 나갈 부분이 `Output`인터페이스를 구현한다면 돌아갈 듯해

입력은 안드로이드 폰에서, 출력은 내 텍스트파일로, 
이러면 `AndroidInputFileOutput implements IO` 그리고 각각 환경에 맞게 구현하는거지

이미 텍스트출력하는 부분이 구현이 되어 있어서 `FileOutput implements Ouput`클래스가 존재한다면, 
`AndroidInputFileOutput`을 작성할 때 `FileOutput`은 필드로 들고 위임을 해주면 되고

lette1394 replied:
de9e847 3ba899e eadc0da 참고하면 좋을 듯해

JuhoKang reviewed (commented):
일단 중간에서 한번 끊겠습니다. `RenderPolicy`부터 다시볼게요

lette1394 committed:
Add Test for Main using FileInput

Input에만 파일을 사용, Output은 그대로 표준 출력

lette1394 committed:
Add dummy data for FileInput test

lette1394 committed:
Add implements for FileInputTest

File을 핸들링 하기 때문에, close가 필요하다.
java.io.Closable은 확인된 예외를 던지므로, 새로운 인터페이스 생성.

편의를 위해 ClosableIO 까지 생성함.
